File: resources/android/DatePickerRenderer.kt
package dev.firstlightui.plugins.firstlight_ui.ui

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import com.nativephp.mobile.ui.nativerender.NativeUIBridge
import com.nativephp.mobile.ui.nativerender.NativeUINode
import com.nativephp.mobile.ui.nativerender.NativeUITree
import com.nativephp.plugins.native_ui.NativeUITheme
import java.time.LocalDate
import java.time.ZoneId

sealed interface DatePickerRendererEvent {
    val wireName: String

    data class Change(val callbackId: Int, val nodeId: Int, val value: String) : DatePickerRendererEvent {
        override val wireName = "SELECT_CHANGE"
    }
}

class DatePickerRendererEvents(private val send: (DatePickerRendererEvent) -> Unit) {
    fun dispatch(event: DatePickerRendererEvent) = send(event)

    companion object {
        val native = DatePickerRendererEvents { event ->
            when (event) {
                is DatePickerRendererEvent.Change -> NativeUIBridge.sendSelectChangeEvent(
                    event.callbackId,
                    event.nodeId,
                    event.value,
                )
            }
        }
    }
}

data class DatePickerRendererConfiguration(
    val nodeId: Int,
    val hasValue: Boolean,
    val value: String,
    val min: String,
    val max: String,
    val label: String,
    val placeholder: String,
    val helper: String,
    val error: String,
    val required: Boolean,
    val disabled: Boolean,
    val locale: String,
    val timezone: String,
    val confirmLabel: String,
    val cancelLabel: String,
    val accessibilityLabel: String,
    val accessibilityHint: String,
    val onChangeCallback: Int,
) {
    constructor(node: NativeUINode) : this(
        nodeId = node.id,
        hasValue = node.props.getBool("has_value"),
        value = node.props.getString("value"),
        min = node.props.getString("min"),
        max = node.props.getString("max"),
        label = node.props.getString("label"),
        placeholder = node.props.getString("placeholder"),
        helper = node.props.getString("helper"),
        error = node.props.getString("error"),
        required = node.props.getBool("required"),
        disabled = node.props.getBool("disabled"),
        locale = node.props.getString("locale"),
        timezone = node.props.getString("timezone"),
        confirmLabel = node.props.getString("confirm_label").ifEmpty { "Confirm" },
        cancelLabel = node.props.getString("cancel_label").ifEmpty { "Cancel" },
        accessibilityLabel = node.props.getString("a11y_label").ifEmpty { node.props.getString("label") },
        accessibilityHint = node.props.getString("a11y_hint"),
        onChangeCallback = node.props.getCallbackId("on_change"),
    )

    val acceptedValue: String? get() = value.takeIf { hasValue }
    val presentationFingerprint: List<Any>
        get() = listOf(hasValue, value, min, max, locale, timezone, disabled)
}

class DatePickerRendererState(node: NativeUINode) {
    var configuration by mutableStateOf(DatePickerRendererConfiguration(node)); private set
    var draft by mutableStateOf<String?>(null); private set
    var isPresented by mutableStateOf(false); private set
    var presentationVersion by mutableIntStateOf(0); private set

    fun open(today: LocalDate = localToday(configuration.timezone)): Boolean {
        if (configuration.disabled || configuration.onChangeCallback == 0) return false

        draft = clamp(configuration.acceptedValue ?: today.toString())
        presentationVersion += 1
        isPresented = true
        return true
    }

    fun userSelected(value: String) {
        if (isPresented) draft = clamp(value)
    }

    fun cancel() {
        draft = null
        isPresented = false
    }

    fun confirm(): DatePickerRendererEvent? {
        val selected = draft
        val published = configuration
        cancel()

        if (selected == null || published.disabled || published.onChangeCallback == 0) return null
        if (selected == published.acceptedValue) return null

        return DatePickerRendererEvent.Change(
            callbackId = published.onChangeCallback,
            nodeId = published.nodeId,
            value = selected,
        )
    }

    fun serverPublished(tree: NativeUITree) {
        val node = tree.root.findDatePickerNode(configuration.nodeId) ?: return
        publish(DatePickerRendererConfiguration(node))
    }

    fun serverPublished(node: NativeUINode) {
        if (node.id == configuration.nodeId) publish(DatePickerRendererConfiguration(node))
    }

    private fun publish(published: DatePickerRendererConfiguration) {
        val presentationChanged = published.presentationFingerprint != configuration.presentationFingerprint
        configuration = published
        if (isPresented && presentationChanged) cancel()
    }

    private fun clamp(value: String): String = value
        .let { candidate -> configuration.min.takeIf(String::isNotEmpty)?.let { maxOf(candidate, it) } ?: candidate }
        .let { candidate -> configuration.max.takeIf(String::isNotEmpty)?.let { minOf(candidate, it) } ?: candidate }

    companion object {
        fun localToday(timezone: String): LocalDate = LocalDate.now(
            timezone.takeIf(String::isNotEmpty)?.let(ZoneId::of) ?: ZoneId.systemDefault(),
        )
    }
}

object FirstlightDatePickerRenderer {
    @Composable
    fun Render(
        node: NativeUINode,
        modifier: Modifier,
        events: DatePickerRendererEvents = DatePickerRendererEvents.native,
    ) {
        val state = remember(node.id) { DatePickerRendererState(node) }
        LaunchedEffect(node.props, node.onPress) { state.serverPublished(node) }
        NativeUIBridge.currentTree.value?.let { tree ->
            LaunchedEffect(tree) { state.serverPublished(tree) }
        }

        FirstlightDatePickerControl(
            state = state,
            tokens = if (isSystemInDarkTheme()) NativeUITheme.dark else NativeUITheme.light,
            onOpen = state::open,
            onCancel = state::cancel,
            onSelect = state::userSelected,
            onConfirm = { state.confirm()?.let(events::dispatch) },
            modifier = modifier,
        )
    }
}

private fun NativeUINode.findDatePickerNode(id: Int): NativeUINode? {
    if (this.id == id) return this
    children.forEach { child -> child.findDatePickerNode(id)?.let { return it } }
    return null
}

File: resources/ios/DatePickerRenderer.swift
import SwiftUI

enum DatePickerRendererEvent: Equatable {
    case change(callbackId: Int, nodeId: Int, value: String)

    var wireName: String { "SELECT_CHANGE" }
}

struct DatePickerRendererEvents {
    let send: @MainActor (DatePickerRendererEvent) -> Void

    static let native = DatePickerRendererEvents { event in
        switch event {
        case let .change(callbackId, nodeId, value):
            NativeUIBridge.sendSelectChangeEvent(callbackId, nodeId: nodeId, value: value)
        }
    }
}

struct DatePickerRendererConfiguration: Equatable {
    let nodeID: Int
    let hasValue: Bool
    let value: String
    let minimum: String
    let maximum: String
    let label: String
    let placeholder: String
    let helper: String
    let error: String
    let required: Bool
    let disabled: Bool
    let locale: String
    let timezone: String
    let confirmLabel: String
    let cancelLabel: String
    let accessibilityLabel: String
    let accessibilityHint: String
    let onChangeCallback: Int

    init(node: NativeUINode) {
        let props = node.props
        nodeID = node.id
        hasValue = props.getBool("has_value")
        value = props.getString("value")
        minimum = props.getString("min")
        maximum = props.getString("max")
        label = props.getString("label")
        placeholder = props.getString("placeholder")
        helper = props.getString("helper")
        error = props.getString("error")
        required = props.getBool("required")
        disabled = props.getBool("disabled")
        locale = props.getString("locale")
        timezone = props.getString("timezone")
        confirmLabel = props.getString("confirm_label", default: "Confirm")
        cancelLabel = props.getString("cancel_label", default: "Cancel")
        let explicitLabel = props.getString("a11y_label")
        accessibilityLabel = explicitLabel.isEmpty ? label : explicitLabel
        accessibilityHint = props.getString("a11y_hint")
        onChangeCallback = props.getCallbackId("on_change")
    }

    var acceptedValue: String? { hasValue ? value : nil }
    var isInteractive: Bool { !disabled && onChangeCallback != 0 }

    fileprivate var presentationFingerprint: String {
        [hasValue ? "1" : "0", value, minimum, maximum, locale, timezone, disabled ? "1" : "0"]
            .joined(separator: "\u{1F}")
    }
}

struct DatePickerRendererState {
    var configuration: DatePickerRendererConfiguration
    var draft: String?
    var isPresented = false
    var presentationVersion = 0

    init(node: NativeUINode) {
        configuration = DatePickerRendererConfiguration(node: node)
        draft = nil
    }

    @discardableResult
    mutating func open(today: String? = nil) -> Bool {
        guard configuration.isInteractive else { return false }
        let seed = configuration.acceptedValue
            ?? today
            ?? DatePickerCalendar.today(timezone: configuration.timezone)
        draft = clamp(seed)
        presentationVersion += 1
        isPresented = true
        return true
    }

    mutating func userSelected(_ value: String) {
        guard isPresented else { return }
        draft = clamp(value)
    }

    mutating func cancel() {
        draft = nil
        isPresented = false
    }

    mutating func confirm() -> DatePickerRendererEvent? {
        let selected = draft
        let published = configuration
        cancel()

        guard let selected, published.isInteractive, selected != published.acceptedValue else {
            return nil
        }

        return .change(
            callbackId: published.onChangeCallback,
            nodeId: published.nodeID,
            value: selected
        )
    }

    mutating func serverPublished(_ tree: NativeUITree) {
        guard let node = Self.findNode(id: configuration.nodeID, in: tree.root) else { return }
        publish(DatePickerRendererConfiguration(node: node))
    }

    private mutating func publish(_ published: DatePickerRendererConfiguration) {
        let presentationChanged = published.presentationFingerprint != configuration.presentationFingerprint
        configuration = published
        if isPresented && presentationChanged { cancel() }
    }

    private func clamp(_ value: String) -> String {
        let aboveMinimum = configuration.minimum.isEmpty ? value : Swift.max(value, configuration.minimum)
        return configuration.maximum.isEmpty ? aboveMinimum : Swift.min(aboveMinimum, configuration.maximum)
    }

    private static func findNode(id: Int, in node: NativeUINode) -> NativeUINode? {
        if node.id == id { return node }
        for child in node.children {
            if let found = findNode(id: id, in: child) { return found }
        }
        return nil
    }
}

enum DatePickerCalendar {
    static func calendar(timezone: String) -> Calendar {
        var calendar = Calendar(identifier: .gregorian)
        calendar.locale = Locale(identifier: "en_US_POSIX")
        calendar.timeZone = TimeZone(identifier: timezone) ?? .current
        return calendar
    }

    static func date(from canonical: String, timezone: String) -> Date {
        let parts = canonical.split(separator: "-").compactMap { Int($0) }
        precondition(parts.count == 3, "Date Picker received a noncanonical date")
        let value = calendar(timezone: timezone).date(
            from: DateComponents(year: parts[0], month: parts[1], day: parts[2])
        )
        return value ?? Date(timeIntervalSince1970: 0)
    }

    static func canonical(from date: Date, timezone: String) -> String {
        let components = calendar(timezone: timezone).dateComponents([.year, .month, .day], from: date)
        return String(format: "%04d-%02d-%02d", components.year!, components.month!, components.day!)
    }

    static func today(timezone: String, now: Date = Date()) -> String {
        canonical(from: now, timezone: timezone)
    }

    static func display(_ canonical: String, locale: String, timezone: String) -> String {
        let formatter = DateFormatter()
        formatter.calendar = calendar(timezone: timezone)
        formatter.timeZone = formatter.calendar.timeZone
        formatter.locale = locale.isEmpty ? .current : Locale(identifier: locale)
        formatter.dateStyle = .medium
        formatter.timeStyle = .none
        return formatter.string(from: date(from: canonical, timezone: timezone))
    }
}

struct DatePickerRenderer: View {
    @ObservedObject private var themeStore = NativeUITheme.shared
    @ObservedObject private var bridge = NativeUIBridge.shared
    @Environment(\.colorScheme) private var colorScheme
    @State private var state: DatePickerRendererState
    private let events: DatePickerRendererEvents

    init(node: NativeUINode, events: DatePickerRendererEvents = .native) {
        _state = State(initialValue: DatePickerRendererState(node: node))
        self.events = events
    }

    var body: some View {
        FirstlightDatePickerControl(
            state: $state,
            tokens: themeStore.resolve(for: colorScheme),
            onConfirm: {
                if let event = state.confirm() { events.send(event) }
            }
        )
        .onReceive(bridge.$currentTree.compactMap { $0 }) { tree in
            state.serverPublished(tree)
        }
    }
}
